packages/worker/src/worker.ts

```diff
@@ -993,260 +999,246 @@ export class Worker {
+          );
+        }
+      }
+
+      let isFatalError = false;
+      try {
+        const decodedActivation = await this.workflowCodecRunner.decodeActivation(activation);
+        const unencodedCompletion = await workflow.workflow.activate(decodedActivation);
+        const completion = await this.workflowCodecRunner.encodeCompletion(unencodedCompletion);
+        this.log.trace('Completed activation', workflow.logAttributes);
+
+        return { state: workflow, output: { close, completion } };
+      } catch (err) {
+        if (err instanceof UnexpectedError) {
+          isFatalError = true;
+        }
+        throw err;
+      } finally {
+        // Fatal error means we cannot call into this workflow again unfortunately
+        if (!isFatalError) {
+          // When processing workflows through runReplayHistories, Core may still send non-replay
+          // activations on the very last Workflow Task in some cases. Though Core is technically exact
+          // here, the fact that sinks marked with callDuringReplay = false may get called on a replay
+          // worker is definitely a surprising behavior. For that reason, we extend the isReplaying flag in
+          // this case to also include anything running under in a replay worker.
+          const isReplaying = activation.isReplaying || this.isReplayWorker;
+
+          const calls = await workflow.workflow.getAndResetSinkCalls();
+          await this.processSinkCalls(calls, isReplaying);
+        }
+      }
+    } catch (error) {
+      if (error instanceof UnexpectedError) {
+        // rethrow and fail the worker
+        throw error;
+      }
+      this.log.error('Failed to activate workflow', {
+        ...(workflow ? workflow.logAttributes : { runId: activation.runId }),
+        error,
+        workflowExists: workflow !== undefined,
+      });
+      const completion = coresdk.workflow_completion.WorkflowActivationCompletion.encodeDelimited({
+        runId: activation.runId,
+        failed: {
+          failure: await encodeErrorToFailure(this.options.loadedDataConverter, error),
+        },
+      }).finish();
+      // We do not dispose of the Workflow yet, wait to be evicted from Core.
+      // This is done to simplify the Workflow lifecycle so Core is the sole driver.
+      return { state: undefined, output: { close: true, completion } };
+    }
+  }
+
+  protected async createWorkflow(
+    activation: coresdk.workflow_activation.WorkflowActivation,
+    startWorkflow: coresdk.workflow_activation.IStartWorkflow
+  ): Promise<WorkflowWithLogAttributes> {
+    const workflowCreator = this.workflowCreator!;
+    if (
+      !(
+        startWorkflow.workflowId != null &&
+        startWorkflow.workflowType != null &&
+        startWorkflow.randomnessSeed != null &&
+        startWorkflow.firstExecutionRunId != null &&
+        startWorkflow.attempt != null &&
+        startWorkflow.startTime != null
+      )
```

Comment:
Can invert this whole condition.

Reply:
Also here, I vote we make deMorgan happy in a follow-on. In the PR so far I basically haven't changed anything other than the restructuring.